File: dijkstra.cpp
#include <cstdio>
#include <vector>
#include <set>

constexpr int inf = 2e9 + 1;

int main()
{
    int n = 0;
    std::scanf("%d", &n);
	
    std::vector<std::vector<std::pair<int, int>>> gr(n);
    std::vector<int> dist(n, inf);
    std::vector<int> pred(n, -1);
    std::set<std::pair<int, int>> s; // dist, v

	for (int i = 0; i < n; i++)
	{
        int x = 0, y = 0, wt = 0;
        std::scanf("%d%d%d", &x, &y, &wt), x--, y--;

        gr[x].push_back({ y, wt });
		// gr[y].push_back(x); - для неориетированного графа
	}

    int start = 0, finish = 0;
    std::scanf("%d%d", &start, &finish), start--, finish--;

    s.insert({0, start});
    dist[start] = 0;

	while (!s.empty())
	{
        const int v = s.begin()->second;
        s.erase(s.begin());

		for (auto p : gr[v])
		{
            const int u = p.first;
            const int wt = p.second;

			if (dist[v] + wt < dist[u])
			{
                s.erase({ dist[u], u });
                dist[u] = dist[v] + wt;
                s.insert({ dist[u], u });
                pred[u] = v;
			}
		}
	}

    std::vector<int> ans;
    ans.push_back(finish);
	
    int prev = pred[finish];

	while (prev != -1)
	{
        ans.push_back(prev);
        prev = pred[prev];
	}

    std::reverse(ans.begin(), ans.end());

    for (int i : ans)
        std::printf("%d ", i);

    return 0;
}

File: dijkstra.kt
import java.util.*

const val INF = 2e9.toInt()

internal object Cmp : Comparator<Pair<Int, Int>> {
    override fun compare(s1: Pair<Int, Int>, s2: Pair<Int, Int>): Int {
        if (s1.first != s2.first)
            return s1.first.compareTo(s2.first)
        return s1.second.compareTo(s2.second)
    }
}

fun main() {
    val n = (readLine() ?: return).toInt()
    val price = (readLine() ?: return).split(" ").map { x -> x.toInt() }
    val m = (readLine() ?: return).toInt()
    val gr = MutableList(n) { MutableList(0) { 0 to 0 } } // v / wt

    for (i in 0 until m) {
        val inp = (readLine() ?: return).split(' ').map { x -> x.toInt() }
        val v = inp.first() - 1
        val u = inp.last() - 1

        gr[v].add(u to price[v])
        gr[u].add(v to price[u])
    }

    val prev = MutableList(n) { -1 }
    val dist = MutableList(n) { INF }
    val se = TreeSet(Cmp)

    se.add(0 to 0)
    dist[0] = 0

    while (!se.isEmpty()) {
        val v = se.first().second
        se.remove(se.first())

        for ((u, wt) in gr[v]) {
            if (dist[v] + wt < dist[u]) {
                se.remove(dist[u] to u)
                dist[u] = dist[v] + wt
                se.add(dist[u] to u)
                prev[u] = v
            }
        }
    }

    println({ if (dist.last() == INF) -1 else dist.last() }())
}
